src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SubSourceLogic.cs

```diff
@@ -147,23 +201,68 @@ protected void ConfigureSubscription(Action<IImmutableSet<TopicPartition>> parti
-                    partitionsRevokedCb));
+                return new PartitionEventHandlers.Chain(
+                    subscription.PartitionEventsHandler.GetOrElse(PartitionEventHandlers.Empty.Instance),
+                    new PartitionEventHandlers.AsyncCallbacks(subscription, SourceActor.Ref, partitionsAssignedCb,
+                        partitionsRevokedCb));
             }
         }
-        
+
+        private class FlushMessagesOfRevokedPartitionsHandler : IPartitionEventHandler
+        {
+            private IImmutableSet<TopicPartitionOffset> _lastRevoked = ImmutableHashSet<TopicPartitionOffset>.Empty;
+            private readonly SubSourceLogic<K, V, TMessage> _stageLogic;
+
+            public FlushMessagesOfRevokedPartitionsHandler(SubSourceLogic<K, V, TMessage> stageLogic)
+            {
+                _stageLogic = stageLogic;
+            }
+
+            public void OnRevoke(IImmutableSet<TopicPartitionOffset> revokedTopicPartitions,
+                IRestrictedConsumer consumer)
+            {
+                _lastRevoked = revokedTopicPartitions;
+            }
+
+            public void OnLost(IImmutableSet<TopicPartitionOffset> revokedTopicPartitions, IRestrictedConsumer consumer)
+            {
+                foreach (var tp in revokedTopicPartitions)
+                {
+                    if (_stageLogic._subSources.TryGetValue(tp.TopicPartition, out var control))
+                        control.FilterRevokedPartitionsCb(revokedTopicPartitions);
+                }
+            }
+
+            public void OnAssign(IImmutableSet<TopicPartition> assignedTopicPartitions, IRestrictedConsumer consumer)
+            {
+                // remove all of our previous revoked partitions that are not in the new assignment
+                var safeToRevoke = _lastRevoked
+                    .Where(c => !assignedTopicPartitions.Contains(c.TopicPartition))
+                    .ToImmutableHashSet();
+
+                foreach (var tp in safeToRevoke)
+                {
+                    if(_stageLogic._subSources.TryGetValue(tp.TopicPartition, out var control))
+                        control.FilterRevokedPartitionsCb(safeToRevoke);
+                }
+            }
+
+            public void OnStop(IImmutableSet<TopicPartition> topicPartitions, IRestrictedConsumer consumer){}
+        }
+
         /// <summary>
         /// Opportunity for subclasses to add their logic to the partition assignment callbacks.
         /// </summary>
         protected virtual IPartitionEventHandler AddToPartitionAssignmentHandler(IPartitionEventHandler handler)
         {
-            return handler;
+            return new PartitionEventHandlers.Chain(handler, new FlushMessagesOfRevokedPartitionsHandler(this));
```

Comment:
Chains the `FlushMessagesOfRevokedPartitionsHandler` to the other two `IPartitionEventHandler`s

Reply:
The chaining needs to be moved to where `AddToPartitionAssignmentHandler` is called, the `AddToPartitionAssignmentHandler` method is supposed to be empty.